### PR TITLE
Added in new SNR method

### PR DESCRIPTION
**Pull request by brechmos-stsci**
New SNR method was requested by @nluetzge.

Fixes https://github.com/astropy/specutils/issues/391

### CODE REVIEW

**Comment by brechmos-stsci** [comment]
cc @nluetzge

File: docs/analysis.rst [hunk 1]
```diff
@@ -47,6 +47,20 @@ spectrum:
     <Quantity 9.84136331>
 
 
+A second method to calculate SNR does not require the uncertainty defined
+on the `~specutils.Spectrum1D` object. This computes the signal to noise 
+ratio DER_SNR following the definition set forth by the Spectral 
+Container Working Group of ST-ECF, MAST and CADC. This is based on the
```
Comment (eteq) [inline docs/analysis.rst]: Can you add a link to the article about this (https://www.spacetelescope.org/static/archives/stecfnewsletters/pdf/hst_stecf_0042.pdf) somewhere? Right now it seems like it's an algorithm "out-of-nowhere", but that article pretty clearly outlines the motivation.

Maybe even pull a one-sentence (at most) description of the algorithm and/or its tradeoffs, derived from what's in that article?  (but maybe that's too hard to do, in which case just a reference to the article describing it is sufficient.)

File: specutils/analysis/uncertainty.py [hunk 2]
```diff
@@ -89,3 +89,92 @@ def _snr_single_region(spectrum, region=None):
     # the axis=-1 will enable this to run on single-dispersion, single-flux
     # and single-dispersion, multiple-flux
     return np.mean(flux / uncertainty, axis=-1)
+
+
+def snr_der(spectrum, region=None):
```
Comment (eteq) [inline specutils/analysis/uncertainty.py]: I found the `der`pretty confusing here.  What about `snr_derived` (which is what the "der" seem to mean here) or `snr_nouncertainty`?  @nluetzge, do you have any opinion on that?

File: specutils/analysis/uncertainty.py [hunk 3]
```diff
@@ -89,3 +89,92 @@ def _snr_single_region(spectrum, region=None):
...
+
+    flux = calc_spectrum.flux
+
+    # Values that are exactly zero (padded) are skipped
+    n = len(flux)
+
+    # For spectra shorter than this, no value can be returned
+    if (n>4):
+        signal = np.median(flux)
+        noise  = 0.6052697 * np.median(np.abs(2.0 * flux[2:n-2] - flux[0:n-4] - flux[4:n]))
```
Comment (eteq) [inline specutils/analysis/uncertainty.py]: it's kind of trivial, but what about using the `astropy.stats` MAD implementation? http://docs.astropy.org/en/stable/api/astropy.stats.median_absolute_deviation.html - the main advantage there is that it treats NaN's a bit more reasonably if you choose `ignore_nan=True` (which could either be a keyword added or just make it True, which I think is my preference)

File: specutils/analysis/uncertainty.py (hunk 2 above)
Comment (brechmos-stsci) [reply specutils/analysis/uncertainty.py]: I would stay away from `snr_nouncertainty` as there might be other algorithms that don't use the uncertainty. I like the  `snr_derived`, personally, as that is what it talked about in the PDF.

Comment (hcferguson) [reply specutils/analysis/uncertainty.py]: How about  `snr_robust_estimate`?

**Comment by gderosa2004** [comment]
@eteq @nluetzge I went through the implementation, and I think that the conditions for this implementation to be an unbiased estimator of the SNR are kind of strict. In particular:
               * the noise is uncorrelated in wavelength bins spaced two pixels apart -- this is not going to happen if the spectrum is not rebinned at least over the resolution elements. 
               * for large wavelength regions, the signal over the scale of 5 or
	         more pixels can be approximated by a straight line -- Not sure if that is easily applicable to sources like QSOs

**Comment by eteq** [comment]
@gderosa2004 - I see your point, but this *is* still a useful somewhat standardizable thing that is used by some archives (apparently), and if I understood the PDF right it's basically "this is only guaranteed to work in certain cases, but actually still works *ok* if you relax the assumptions".

Certainly these caveats should be laid out clearly in the docstrings, though.  One could even imagine writing an explicit check for spectra that have been rebinned (at least by specutils, once we have that implemented) - i.e. it would make a warning "this might be wrong depending on the details of your data"?  Do you think that would be sufficient?

**Comment by hcferguson** [comment]
@gderosa2004 I think the assumption that the *noise* is uncorrelated in wavelength bins spaced two pixels apart is distinct from the correlation imposed by the spectral resolution. The spectral resolution creates a correlation of the *signal* not the noise. The noise is usually Poisson noise, detector dark count and readout noise (irrelevant for COS). I think these are generally uncorrelated even between adjacent pixels, but drizzling or other types of co-addition can induce a correlation, which is why it's wise to skip a pixel.

**Comment by eteq** [comment]
(This discussion I think highlights why it should be written out in the docs! :wink:)

**Comment by brechmos-stsci** [comment]
@eteq or @hcferguson I'll let you guys wordsmith the docs for the `snr_derived`. I judiciously copied in some words from @gderosa2004, above, regarding the independence of the data for using this algorithm.  Otherwise, when you feel it is ready, if you do, feel free to hit merge.

**Comment by brechmos-stsci** [comment]
@eteq Feel free to merge unless you have some other things you want me to change. Not sure if you would like to make changes to the documentation?

**Comment by brechmos-stsci** [comment]
@eteq Are you good with this one and can we merge?

**Comment by nmearl** [comment]
@eteq Are you satisfied with this PR?

**Comment by bsipocz** [comment]
This PR is another good example that it's good practice to rebase old PRs before merging to make sure CI is still all green (the same issue bite us a few times in astropy, so it's non trivial to change the practice).

**Pull request merged**